File: ketch-android/Android Ketch Pref Center using Java/app/src/main/java/com/ketch/sample/webview/KetchWebView.java
package com.ketch.sample.webview;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.webkit.ConsoleMessage;
import android.webkit.JavascriptInterface;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.webkit.WebViewAssetLoader;
import androidx.webkit.WebViewClientCompat;

import java.util.List;

public class KetchWebView extends WebView {
    private static final String TAG = KetchWebView.class.getSimpleName();

    private static final String WILL_NOT_SHOW = "willNotShow";
    private static final String CLOSE = "close";
    private static final String SET_CONSENT = "setConsent";

    private KetchListener listener;

    private String ketchUrl;

    public KetchWebView(@NonNull Context context) {
        this(context, null);
    }

    public KetchWebView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public KetchWebView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    @SuppressLint("SetJavaScriptEnabled")
    public KetchWebView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        setVisibility(View.GONE);

        getSettings().setJavaScriptEnabled(true);

        WebViewAssetLoader assetLoader = new WebViewAssetLoader.Builder()
                .addPathHandler("/assets/", new WebViewAssetLoader.AssetsPathHandler(context))
                .addPathHandler("/res/", new WebViewAssetLoader.ResourcesPathHandler(context))
                .build();

        setWebViewClient(new LocalContentWebViewClient(assetLoader));
        setWebContentsDebuggingEnabled(true);

        addJavascriptInterface(
                new PreferenceCenterJavascriptInterface(this),
                "androidListener"
        );

        //receive console messages from the WebView
        WebChromeClient webChromeClient = new WebChromeClient() {
            @Override
            public boolean onConsoleMessage(ConsoleMessage consoleMessage) {
                Log.d(TAG, consoleMessage.message());
                return true;
            }
        };
    }

    public void setListener(KetchListener listener) {
        this.listener = listener;
    }

    void init(String orgCode, String property, List<Identity> identities) {
        //pass in the property code and  to be used with the Ketch Smart Tag
        StringBuilder url = new StringBuilder();
        url.append(String.format("https://appassets.androidplatform.net/assets/index.html?orgCode=%1$s&propertyName=%2$s", orgCode, property));

        for (Identity identity : identities) {
            url.append(String.format("&%1$s=%2$s", identity.getCode(), identity.getValue()));
        }

        ketchUrl = url.toString();

        loadUrl(ketchUrl);
    }

    private static class LocalContentWebViewClient extends WebViewClientCompat {
        private final WebViewAssetLoader assetLoader;

        private LocalContentWebViewClient(WebViewAssetLoader assetLoader) {
            this.assetLoader = assetLoader;
        }

        @Override
        @Nullable
        public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
            return assetLoader.shouldInterceptRequest(request.getUrl());
        }
    }

    public void show() {
        loadUrl("about:blank");
        clearHistory();
        loadUrl(ketchUrl);
        setVisibility(View.VISIBLE);
    }

    private static class PreferenceCenterJavascriptInterface {
        private final KetchWebView ketchWebView;

        private PreferenceCenterJavascriptInterface(KetchWebView ketchWebView) {
            this.ketchWebView = ketchWebView;
        }

        @JavascriptInterface
        public void hideExperience(@Nullable String status) {
            Log.d(TAG, String.format("hideExperience: %s", status));
            if (status != null && (status.equalsIgnoreCase(SET_CONSENT) || (status.equalsIgnoreCase(CLOSE)))) {
                runOnMainThread(() -> {
                    if (ketchWebView.listener != null) {
                        ketchWebView.listener.onClose();
                    }
                    ketchWebView.setVisibility(View.GONE);
                });
            }
        }

        @JavascriptInterface
        public void usprivacy_updated(@Nullable String ccpaString) {
            Log.d(TAG, String.format("onCCPAUpdate: %s", ccpaString));
            runOnMainThread(() -> {
                if (ketchWebView.listener != null) {
                    ketchWebView.listener.onCCPAUpdate(ccpaString);
                }
            });
        }

        @JavascriptInterface
        public void tcf_updated(@Nullable String tcfString) {
            Log.d(TAG, String.format("onTCFUpdate: %s", tcfString));
            runOnMainThread(() -> {
                ketchWebView.listener.onTCFUpdate(tcfString);
            });
        }

        @JavascriptInterface
        public void environment(@Nullable String environment) {
            Log.d(TAG, String.format("environment: %s", environment));
        }

        @JavascriptInterface
        public void regionInfo(@Nullable String regionInfo) {
            Log.d(TAG, String.format("regionInfo: %s", regionInfo));
        }

        @JavascriptInterface
        public void jurisdiction(@Nullable String jurisdiction) {
            Log.d(TAG, String.format("jurisdiction: %s", jurisdiction));
        }

        @JavascriptInterface
        public void identities(@Nullable String identities) {
            Log.d(TAG, String.format("identities: %s", identities));
        }

        @JavascriptInterface
        public void consent(@Nullable String consent) {
            Log.d(TAG, String.format("consent: %s", consent));
        }

        @JavascriptInterface
        public void showConsentExperience(@Nullable String showConsentExperience) {
            Log.d(TAG, String.format("showConsentExperience: %s", showConsentExperience));
        }

        @JavascriptInterface
        public void showPreferenceExperience(@Nullable String showPreferenceExperience) {
            Log.d(TAG, String.format("showPreferenceExperience: %s", showPreferenceExperience));
        }

        private void runOnMainThread(Runnable runnable) {
            new Handler(Looper.getMainLooper()).post(runnable);
        }
    }

    interface KetchListener {
        void onCCPAUpdate(@Nullable String ccpaString);

        void onTCFUpdate(@Nullable String tcfString);

        void onClose();
    }
}

File: ketch-android/Android Ketch Pref Center using Kotlin/app/src/main/java/com/ketch/sample/pref/KetchWebView.kt
package com.ketch.sample.pref

import android.annotation.SuppressLint
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.util.Log
import android.webkit.ConsoleMessage
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebView
import androidx.core.view.isVisible
import androidx.webkit.WebViewAssetLoader
import androidx.webkit.WebViewClientCompat
import com.google.gson.Gson
import com.google.gson.JsonParseException

@SuppressLint("SetJavaScriptEnabled")
class KetchWebView(context: Context, attrs: AttributeSet?) : WebView(context, attrs) {

    private lateinit var ketchUrl: String
    var listener: KetchListener? = null

    init {
        isVisible = false
        settings.javaScriptEnabled = true
        val assetLoader = WebViewAssetLoader.Builder()
            .addPathHandler("/assets/", WebViewAssetLoader.AssetsPathHandler(context))
            .addPathHandler("/res/", WebViewAssetLoader.ResourcesPathHandler(context))
            .build()

        webViewClient = LocalContentWebViewClient(assetLoader)
        setWebContentsDebuggingEnabled(true)

        addJavascriptInterface(
            PreferenceCenterJavascriptInterface(this),
            "androidListener"
        )

        //receive console messages from the WebView
        webChromeClient = object : WebChromeClient() {
            override fun onConsoleMessage(consoleMessage: ConsoleMessage): Boolean {
                Log.d(TAG, consoleMessage.message())
                return true
            }
        }
    }

    private class LocalContentWebViewClient(private val assetLoader: WebViewAssetLoader) :
        WebViewClientCompat() {
        override fun shouldInterceptRequest(
            view: WebView,
            request: WebResourceRequest
        ): WebResourceResponse? {
            return assetLoader.shouldInterceptRequest(request.url)
        }
    }

    fun init(orgCode: String, property: String, identities: List<Identity> = emptyList()) {
        //pass in the property code and  to be used with the Ketch Smart Tag
        var url =
            "https://appassets.androidplatform.net/assets/index.html?orgCode=$orgCode&propertyName=$property"

        identities.forEach { identity ->
            url = "$url&${identity.code}=${identity.value}"
        }

        ketchUrl = url

        loadUrl(ketchUrl)
    }

    fun show() {
        loadUrl("about:blank")
        clearHistory()
        loadUrl(ketchUrl)
        isVisible = true
    }

    private class PreferenceCenterJavascriptInterface(private val ketchWebView: KetchWebView) {
        @JavascriptInterface
        fun hideExperience(status: String?) {
            Log.d(TAG, "hideExperience: $status")
            if (status?.equals(CLOSE, ignoreCase = true) == true
                || status?.equals(SET_CONSENT, ignoreCase = true) == true
            ) {
                runOnMainThread {
                    ketchWebView.listener?.onClose()
                    ketchWebView.isVisible = false
                }
            }
        }

        @JavascriptInterface
        fun usprivacy_updated(ccpaString: String?) {
            Log.d(TAG, "onCCPAUpdate: $ccpaString")
            runOnMainThread {
                ketchWebView.listener?.onCCPAUpdate(ccpaString)
            }
        }

        @JavascriptInterface
        fun tcf_updated(tcfString: String?) {
            Log.d(TAG, "onTCFUpdate: tcfString: $tcfString")
            runOnMainThread {
                ketchWebView.listener?.onTCFUpdate(tcfString)
            }
        }

        @JavascriptInterface
        fun environment(environment: String?) {
            Log.d(TAG, "environment: $environment")
        }

        @JavascriptInterface
        fun regionInfo(regionInfo: String?) {
            Log.d(TAG, "regionInfo: $regionInfo")
        }

        @JavascriptInterface
        fun jurisdiction(jurisdiction: String?) {
            Log.d(TAG, "jurisdiction: $jurisdiction")
        }

        @JavascriptInterface
        fun identities(identities: String?) {
            Log.d(TAG, "identities: $identities")
        }

        @JavascriptInterface
        fun consent(consentJson: String?) {
            // {"purposes":{"essential_services":true,"tcf.purpose_1":true,"analytics":false,"behavioral_advertising":false,"email_marketing":false,"data_broking":false,"somepurpose_key":false},"vendors":[]}
            val consent = try {
                Gson().fromJson(consentJson, Consent::class.java)
            } catch (ex: JsonParseException) {
                Log.e(TAG, ex.message, ex)
            }
            Log.d(TAG, "consent: $consent")
        }

        @JavascriptInterface
        fun willShowExperience(willShowExperience: String?) {
            Log.d(TAG, "willShowExperience: $willShowExperience")
        }

        @JavascriptInterface
        fun showConsentExperience(showConsentExperience: String?) {
            Log.d(TAG, "showConsentExperience: $showConsentExperience")
        }

        @JavascriptInterface
        fun showPreferenceExperience(showPreferenceExperience: String?) {
            Log.d(TAG, "showPreferenceExperience: $showPreferenceExperience")
        }

        private fun runOnMainThread(action: () -> Unit) {
            Handler(Looper.getMainLooper()).post {
                action.invoke()
            }
        }
    }

    interface KetchListener {
        fun onCCPAUpdate(ccpaString: String?)
        fun onTCFUpdate(tcfString: String?)
        fun onClose()
    }

    companion object {
        private val TAG: String = KetchWebView::class.java.simpleName

        private const val WILL_NOT_SHOW = "willNotShow"
        private const val CLOSE = "close"
        private const val SET_CONSENT = "setConsent"
    }
}
